server/src/main/java/com/cloud/hypervisor/kvm/discoverer/LibvirtServerDiscoverer.java

```diff
@@ -259,9 +259,17 @@ private void setupAgentSecurity(final Connection sshConnection, final String age
             sshConnection = new Connection(agentIp, 22);
 
             sshConnection.connect(null, 60000, 60000);
-            if (!sshConnection.authenticateWithPassword(username, password)) {
-                s_logger.debug("Failed to authenticate");
-                throw new DiscoveredWithErrorException("Authentication error");
+
+            final String privateKey = _configDao.getValue("ssh.privatekey");
```

Comment:
@weizhouapache I think, it is better to try with ssh key or password (based on the authentication mode), not with other when one of it fails. You can determine authentication mode, either from the user input or may be from password/privatekey config.

Reply:
@sureshanaparti 
isn't it better to support both ? hosts are still manageable if one of the authentication fails.

Reply:
> 

if one authentication fails, try other may not be good.

Reply:
> if one authentication fails, try other may not be good.

@sureshanaparti ok. let's ask other guys.

@rhtyd @DaanHoogland what's your opinion ?